File: Programmers/[87391] 공 이동 시뮬레이션/ONE/87391.java
class Solution {
    public long solution(int n, int m, int x, int y, int[][] queries) {
        int len = queries.length;
        long rowStart = x, rowEnd = x;
        long colStart = y, colEnd = y;

        for (int i = len - 1; i >= 0; i--) {
            int direction = queries[i][0];
            int distance = queries[i][1];

            switch (direction) {
                case 0: // 열을 증가
                    if (colStart != 0)
                        colStart += distance;
                    colEnd += distance;
                    if (colEnd > m - 1)
                        colEnd = m - 1;
                    break;
                case 1: // 열을 감소
                    colStart -= distance;
                    if (colStart < 0)
                        colStart = 0;
                    if (colEnd != m - 1)
                        colEnd -= distance;
                    break;
                case 2: // 행을 증가
                    if (rowStart != 0)
                        rowStart += distance;
                    rowEnd += distance;
                    if (rowEnd > n - 1)
                        rowEnd = n - 1;
                    break;
                case 3: // 행을 감소
                    rowStart -= distance;
                    if (rowStart < 0)
                        rowStart = 0;
                    if (rowEnd != n - 1)
                        rowEnd -= distance;
            }
            if (rowStart > n - 1 || rowEnd < 0 || colStart > m - 1 || colEnd < 0)
                return 0;
        }
        return (rowEnd - rowStart + 1) * (colEnd - colStart + 1);
    }
}

File: Programmers/[87391] 공 이동 시뮬레이션/Johoseong/87391.py
def solution(n, m, x, y, queries):
    s_r, s_c, e_r, e_c = x, y, x, y
    
    while queries:
        d, dx = queries.pop()
        if d == 0:
            if s_c == 0:
                e_c = min(m - 1, e_c + dx)
            else:
                if s_c + dx >= m: # 불가능 경우
                    return 0
                s_c = min(m - 1, s_c + dx)
                e_c = min(m - 1, e_c + dx)

        elif d == 1:
            if e_c == (m - 1):
                s_c = max(0, s_c - dx)
            else:
                if e_c - dx < 0: # 불가능 경우
                    return 0
                s_c = max(0, s_c - dx)
                e_c = max(0, e_c - dx)

        elif d == 2:
            if s_r == 0:
                e_r = min(n - 1, e_r + dx)
            else:
                if s_r + dx >= n:
                    return 0
                s_r = min(n - 1, s_r + dx)
                e_r = min(n - 1, e_r + dx)
        else:
            if e_r == (n - 1):
                s_r = max(0, s_r - dx)
            else:
                if e_r - dx < 0:
                    return 0
                s_r = max(0, s_r - dx)
                e_r = max(0, e_r - dx)
    
    return (e_r - s_r + 1) * (e_c - s_c + 1)

print(solution(2, 5, 0, 1, [[3, 1], [2, 2], [1, 1], [2, 3], [0, 1], [2, 1]]))